flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/FixedLengthRecordSorter.java

```diff
@@ -447,11 +447,13 @@ public void writeToOutput(final ChannelWriterOutputView output, final int start,
 				num -= recordsPerSegment;
 			} else {
 				// partially filled segment
-				for (; num > 0; num--) {
+				for (; num > 0 && offset <= this.lastEntryOffset; num--, offset += this.recordSize) {
```

Comment:
Is this a bug in the current implementation? If yes, that is critical.

We should pull this into a dedicated fix and add a test in that case.